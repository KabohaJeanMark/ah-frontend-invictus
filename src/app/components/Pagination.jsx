import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

// points for moving left(before) or right(after) to the currentPage
const PAGE_BEFORE = 'BEFORE';
const PAGE_AFTER = 'AFTER';

/**
 * helper function to create a range of numbers.
 * range(1, 5) => [1, 2, 3, 4, 5]
 */
const range = (from, to, step = 1) => {
  let i = from;
  const range = [];
  while (i <= to) {
    range.push(i);
    i += step;
  }

  return range;
}

class Pagination extends Component {
  constructor(props) {
    super(props);
    const { totalRecords = null, pageLimit = 30, pageNeighbours = 0 } = props;

    this.pageLimit = typeof pageLimit === 'number' ? pageLimit : 30;
    this.totalRecords = typeof totalRecords === 'number' ? totalRecords : 0;
    // pageNeighbours can be 0, 1 or 2 or so on.
    this.pageNeighbours = typeof pageNeighbours === 'number'
      ? Math.max(0, Math.min(pageNeighbours, 2))
      : 0;
    // compute total number of pages.
    this.totalPages = Math.ceil(this.totalRecords / this.pageLimit);
    // initialize state.
    this.state = { currentPage: 1 };
  }

  componentDidMount() {
    this.gotoPage(1);
  }

  gotoPage = page => {
    const { onPageChanged = f => f } = this.props;

    const currentPage = Math.max(0, Math.min(page, this.totalPages));

    const paginationData = {
      currentPage,
      totalPages: this.totalPages,
      pageLimit: this.pageLimit,
      totalRecords: this.totalRecords
    };

    this.setState({ currentPage }, () => onPageChanged(paginationData));
  }

  handleClick = page => evt => {
    evt.preventDefault();
    this.gotoPage(page);
  }

  handleMoveLeft = evt => {
    evt.preventDefault();
    this.gotoPage(this.state.currentPage - (this.pageNeighbours * 2) - 1);
  }

  handleMoveRight = evt => {
    evt.preventDefault();
    this.gotoPage(this.state.currentPage + (this.pageNeighbours * 2) + 1);
  }

  /**
   * Given 10 pages and pageNeighbours is set to 2
   * and the currentPage is 5
   * the pagination control will look like this
   *
   * (1)<<{3 4}[5]{6 7 8}>>(10)
   *
   * (x) => terminal pages: first and last, always visible
   * [x] => currentPage
   * {..x} => pageNeighbours
   */
  fetchPageNumbers = () => {
    const totalPages = this.totalPages;
    const currentPage = this.state.currentPage;
    const pageNeighbours = this.pageNeighbours;
    /**
     * Total page numbers to show on the control
     */
    const totalNumbers = (this.pageNeighbours * 2) + 3;
    /**
     * totalNumbers + 2 to cover for the (next) and (prev) controls
     */
    const totalBlocks = totalNumbers + 2;

    if (totalPages > totalBlocks) {
      const startPage = Math.max(2, currentPage - pageNeighbours);
      const endPage = Math.min(totalPages - 1, currentPage + pageNeighbours);
      let pages = range(startPage, endPage);

      /**
       * hasLeftSpill: has hidden pages to the left
       * hasRightSpill: has hidden pages to the right
       * spillOffset: number of hidden pages either to the left or to the right
       */
      const hasLeftSpill = startPage > 2;
      const hasRightSpill = (totalPages - endPage) > 1;
      const spillOffset = totalNumbers - (pages.length + 1);

      switch (true) {
        // handle leftSpill
        case (hasLeftSpill && !hasRightSpill): {
          const extraPages = range(startPage - spillOffset, startPage - 1);
          pages = [PAGE_BEFORE, ...extraPages, ...pages];
          break;
        }
        // handle rightSpill
        case (!hasLeftSpill && hasRightSpill): {
          const extraPages = range(endPage + 1, endPage + spillOffset);
          pages = [...pages, ...extraPages, PAGE_AFTER];
          break;
        }
        // handle currentPage spill
        case (hasLeftSpill && hasRightSpill):
        default: {
          pages = [PAGE_BEFORE, ...pages, PAGE_AFTER];
          break;
        }
      }
      return [1, ...pages, totalPages];
    }
    return range(1, totalPages);
  }

  render() {

    if (!this.totalRecords || this.totalPages === 1) return null;

    const { currentPage } = this.state;
    const pages = this.fetchPageNumbers();

    return (
      <Fragment>
        <div aria-label="articles-pagination">
          <ul className="pagination">
            { pages.map((page, index) => {
              if (page === PAGE_BEFORE) return (
                <li key={index} className="page-item">
                  <a className="page-link" href="#" aria-label="Previous" onClick={this.handleMoveLeft}>
                    <span aria-hidden="true">&laquo;</span>
                    <span className="sr-only">Previous</span>
                  </a>
                </li>
              );
              if (page === PAGE_AFTER) return (
                <li key={index} className="page-item">
                  <a className="page-link" href="#" aria-label="Next" onClick={this.handleMoveRight}>
                    <span aria-hidden="true">&raquo;</span>
                    <span className="sr-only">Next</span>
                  </a>
                </li>
              );
              return (
                <li key={index} className={`page-item${ currentPage === page ? ' active' : ''}`}>
                  <a className="page-link" href="#" onClick={ this.handleClick(page) }>{ page }</a>
                </li>
              );
            }) }
          </ul>
        </div>
      </Fragment>
    );
  }
}

Pagination.propTypes = {
  totalRecords: PropTypes.number.isRequired,
  pageLimit: PropTypes.number,
  pageNeighbours: PropTypes.number,
  onPageChanged: PropTypes.func
};

export default Pagination;
