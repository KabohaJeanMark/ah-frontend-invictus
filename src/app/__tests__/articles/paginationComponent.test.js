import React from 'react';
import ReactDOM from 'react-dom';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import Pagination from '../../components/Pagination';

jest.dontMock('../../components/Pagination');

import ReactTestUtils from 'react-dom/test-utils';

const props = {
  handleClick: jest.fn(),
  handleMoveLeft: jest.fn(),
  handleMoveRight: jest.fn(),
  gotoPage: jest.fn(),
  preventDefault: jest.fn(),
  fetchPageNumbers: jest.fn(),
  totalRecords: 20,
  pageLimit: 1,
  pageNeighbours: 2
};

describe('pagination component', () => {
  let wrapper = shallow(<Pagination {...props} />);

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should handleClick', () => {
    const page = { page: 1 }
    const evt = {
      preventDefault: jest.fn(),
      target: page
    }
    expect(wrapper.instance().handleClick(page)(evt));
  });

  it('should handleMoveLeft', () => {
    const page = { page: 1 }
    const evt = {
      preventDefault: jest.fn(),
      target: page
    }
    expect(wrapper.instance().handleMoveLeft(evt));
  });

  it('should handleMoveRight', () => {
    const page = { page: 2 }
    const evt = {
      preventDefault: jest.fn(),
      target: page
    }
    expect(wrapper.instance().handleMoveRight(evt));
  });
});

describe('test onPageChange', () => {
  it('should display', () => {
    const pagination = ReactTestUtils.renderIntoDocument(
      <Pagination
        totalRecords={12}
        pageLimit={6}
        pageNeighbours={2}
        onPageChanged={props.onPageChanged}
      />
    );
    const previousElement = ReactDOM.findDOMNode(pagination).querySelector('li:first-child');
    const nextElement = ReactDOM.findDOMNode(pagination).querySelector('li:last-child');
    expect(previousElement.className).toBe('page-item active');
    expect(nextElement.className).toBe('page-item');
  });
})
