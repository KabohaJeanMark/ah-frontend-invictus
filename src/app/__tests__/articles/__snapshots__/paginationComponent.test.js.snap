// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination component matches snapshot 1`] = `
<Fragment>
  <div
    aria-label="articles-pagination"
  >
    <ul
      className="pagination"
    >
      <li
        className="page-item active"
        key="0"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          1
        </a>
      </li>
      <li
        className="page-item"
        key="1"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          2
        </a>
      </li>
      <li
        className="page-item"
        key="2"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          3
        </a>
      </li>
      <li
        className="page-item"
        key="3"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          4
        </a>
      </li>
      <li
        className="page-item"
        key="4"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          5
        </a>
      </li>
      <li
        className="page-item"
        key="5"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          6
        </a>
      </li>
      <li
        className="page-item"
        key="6"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          7
        </a>
      </li>
      <li
        className="page-item"
        key="7"
      >
        <a
          aria-label="Next"
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          <span
            aria-hidden="true"
          >
            Â»
          </span>
          <span
            className="sr-only"
          >
            Next
          </span>
        </a>
      </li>
      <li
        className="page-item"
        key="8"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
        >
          20
        </a>
      </li>
    </ul>
  </div>
</Fragment>
`;
